### DISCUSSION TITLE
Possibility to use the same port pair under bidirectional mode

### DISCUSSION THREAD

**Discussion by swg0101** [esnet/iperf#1311] [Ideas]
I am currently trying to track down an issue in which the Linux kernel seems to be dropping packets when an IPv6 flow is offloaded. Interestingly, I could only reproduce this loss when data is being sent and received over the same port pair concurrently. Although iperf3 supports bidirectional testing, two separate flows are opened for the test and it was not possible to mimic the same behavior as an ordinary SIP connection (which uses the same port pair - similar to what packetlosstest.com uses).

How feasible would it be to add something like this to iperf3 to cover this use case?

Thanks!

**Comment by swlars** [comment]
Hi,

Thanks for the question!

Iperf3 was originally designed to only go in one direction, so bidirectional tests were added later. The way it was implemented, Iperf3 uses distinct sockets for different tests, even with bidirectional tests. Changing it would be possible, but it's not a high priority, and it would probably take larger changes.

**Reply by swg0101** [reply]
Understood - having the traffic on the same flow would be helpful in diagnosing flow-based issues, so hopefully, this can be implemented in the future.

**Comment by davidBar-On** [comment]
I have already started to develop a proposed solution - using the same socket for both directions.  I hope that it will be ready in few days.

**Reply by swg0101** [reply]
Oh wow, can't wait to see this :)

**Comment by davidBar-On** [comment]
I have submitter PR #1321 with a proposed enhancement - use the same socket (and therefore same port) for both `bidir` directions.

**Reply by swg0101** [reply]
Thank you for the quick implementation - a quick test on Linux confirming with tcpdump seems to work fine. WIll test on a Windows build next. I don't believe there is any way we can have old servers immediately reject connections made in this way can we?

**Comment by davidBar-On** [comment]
> I don't believe there is any way we can have old servers immediately reject connections made in this way can we?

As far as I know this is not possible.  The server just ignores unknown options received from the client.  There is also no way that the client will check whether the server supports certain options.

**Reply by swg0101** [reply]
Good to know, thank you.
